docs/spec_draft.md

```diff
@@ -3408,6 +3409,117 @@ More formally, `results[:][j0, ..., jR-1] = reduce(input_slices)` where:
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups)`.
+
+Within each `process_group`:
+  * ```
+    reduced_value@process = reduce-all(operand, replica_groups, channel_handle, use_global_device_ids, computation) for each process in process_group
+    ```
+  * ```
+    split_count = dim(process_groups, 1)
+    split_parts@sender = [
+        slice(
+          operand=operand@sender,
+          start_indices=[s0, s1, ..., sR-1],
+            # where
+            #  - sj = 0 if j != split_dimension
+            #  - sj = i * dim(operand) // split_count, if j == split_dimension
+            #  - R = rank(operand)
+          limit_indices=[l0, l1, ..., lR-1],
+            # where
+            #   - lj = dim(operand, j) if j != split_dimension
+            #   - lj = (i + 1) * dim(operand, split_dimension) // split_count, if j == split_dimension
+          strides=[1, ..., 1]
+        ) for i in range(split_count)
+     ]
+     ``` for all `sender` in `process_group`.
+  * `result@receiver = split_parts@sender[receiver_index]` for any sender in process_group,
+      where `receiver_index = index_of(receiver, process_group)`.
+
+### Inputs
+
+| Name                    | Type                                         |
+|-------------------------|----------------------------------------------|
+| `operand`               | tensor of any supported type                 |
+| `scatter_dimension`     | constant of type `si64`                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64` |
+| `channel_id`            | constant of type `si64`                      |
+| `use_global_device_ids` | constant of type `boolean`                   |
+| `computation`           | `function`                                   |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `scatter_dimension` $\in$ [0, rank(`operand`)).
+  * (C2) All values in `replica_groups` are unique.
+  * (C3) `size(replica_groups)` depends on the process grouping strategy:
+    * If `cross_replica`, `num_replicas`.
+    * If `cross_replica_and_partition`, `num_replicas`.
+    * If `flattened_ids`, `num_processes`.
+  * (C4) $0 \le$ `replica_groups[i]` $\lt$ size(`replica_groups`) $\forall i$
+         in `indices(replica_groups)`.
+  * (C5) If `use_global_device_ids = true`, then `channel_id > 0`. [todo](https://github.com/openxla/stablehlo/issues/654)
+  * (C6) `computation` has type `(tensor<E>, tensor<E>) -> (tensor<E>)` where
+         `E = element_type(operand)`.
+  * (C7) `type(result) = type(operand)` except:
+    * `dim(result, scatter_dimension) = dim(operand, scatter_dimension) / dim(process_groups, 1)`.
```

Comment:
* dim(operand, scatter_dimension) % dim(process_groups, 1) == 0

also if this is true we can remove the `//` denoting integer division and assume that division will always be perfect.

Reply:
Thanks! I think the check should be there.